File: server/src/seed.js
// Loads the demo project (converted from the Sibline Port control workbook) and
// creates a starting account. Safe to run more than once: it skips work already done.
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import db from './db.js';
import { hashPassword } from './auth.js';
import { today } from './service.js';

const here = path.dirname(fileURLToPath(import.meta.url));
const seedFile = process.env.SEED_FILE || path.join(here, '..', 'seed', 'sibline-port.json');

const email = process.env.SEED_EMAIL || 'admin@example.com';
const password = process.env.SEED_PASSWORD || 'changeme123';
const name = process.env.SEED_NAME || 'Project Manager';

let user = db.prepare('SELECT * FROM users WHERE email = ?').get(email);
if (!user) {
  const info = db.prepare('INSERT INTO users (email, name, password_hash, role) VALUES (?, ?, ?, ?)')
    .run(email, name, hashPassword(password), 'admin');
  user = db.prepare('SELECT * FROM users WHERE id = ?').get(info.lastInsertRowid);
  console.log(`Created account ${email} (password: ${password})`);
} else {
  console.log(`Account ${email} already exists — leaving it alone`);
}

const seed = JSON.parse(fs.readFileSync(seedFile, 'utf8'));

if (db.prepare('SELECT 1 FROM projects WHERE code = ?').get(seed.project.code)) {
  console.log(`Project ${seed.project.code} already seeded — nothing to do`);
  process.exit(0);
}

const load = db.transaction(() => {
  const p = seed.project;
  const projectId = db.prepare(`
    INSERT INTO projects (code, name, client, description, ntp_date, duration_months,
                          days_per_month, hours_per_month, elapsed_day_offset, status, owner_id)
    VALUES (@code, @name, @client, @description, @ntp_date, @duration_months,
            @days_per_month, @hours_per_month, @elapsed_day_offset, @status, @owner_id)
  `).run({ ...p, owner_id: user.id }).lastInsertRowid;

  const tradeIds = {};
  for (const t of seed.trades) {
    tradeIds[t.key] = db.prepare(
      'INSERT INTO trades (project_id, key, name, budget_hours, color, sort_order) VALUES (?, ?, ?, ?, ?, ?)'
    ).run(projectId, t.key, t.name, t.budget_hours, t.color, t.sort_order).lastInsertRowid;
  }

  const sectionIds = {};
  for (const s of seed.sections) {
    sectionIds[s.code] = db.prepare(
      'INSERT INTO sections (project_id, code, name, sort_order) VALUES (?, ?, ?, ?)'
    ).run(projectId, s.code, s.name, s.sort_order).lastInsertRowid;
  }

  const insertTask = db.prepare(`
    INSERT INTO tasks (project_id, section_id, wbs, name, weight_points, start_month,
                       finish_month, actual_pct, remarks, sort_order)
    VALUES (@project_id, @section_id, @wbs, @name, @weight_points, @start_month,
            @finish_month, @actual_pct, @remarks, @sort_order)
  `);
  const insertAlloc = db.prepare('INSERT INTO task_allocations (task_id, trade_id, pct) VALUES (?, ?, ?)');
  const insertProgress = db.prepare(`
    INSERT INTO progress_updates (task_id, project_id, user_id, previous_pct, actual_pct, note, data_date)
    VALUES (?, ?, ?, 0, ?, 'Imported from control workbook', ?)
  `);

  for (const t of seed.tasks) {
    const taskId = insertTask.run({
      project_id: projectId,
      section_id: sectionIds[t.section_code] ?? null,
      wbs: t.wbs,
      name: t.name,
      weight_points: t.weight_points,
      start_month: t.start_month,
      finish_month: t.finish_month,
      actual_pct: t.actual_pct,
      remarks: t.remarks,
      sort_order: t.sort_order,
    }).lastInsertRowid;

    for (const [key, pct] of Object.entries(t.allocations)) {
      if (pct > 0) insertAlloc.run(taskId, tradeIds[key], pct);
    }
    if (t.actual_pct > 0) {
      insertProgress.run(taskId, projectId, user.id, t.actual_pct, p.ntp_date > today() ? p.ntp_date : today());
    }
  }
  return projectId;
});

const id = load();
console.log(`Seeded "${seed.project.name}" (project #${id}) with ${seed.tasks.length} deliverables ` +
  `across ${seed.sections.length} sections and ${seed.trades.length} trades.`);

File: app/seed.py
"""Loads the demo project (converted from the Sibline Port control workbook)
and creates a starting account. Safe to run more than once: it skips work
already done."""

from __future__ import annotations

import json
import os
from pathlib import Path

from datetime import datetime, timedelta

from .auth import hash_password
from .db import connect, database_path, init_db
from .workflow import default_steps

SEED_FILE = Path(__file__).resolve().parent.parent / "seed" / "sibline-port.json"


def seed(database: str | None = None, seed_file: Path | None = None, quiet: bool = False) -> None:
    path = database or str(database_path())
    init_db(path)
    conn = connect(path)

    def say(message: str) -> None:
        if not quiet:
            print(message)

    try:
        email = os.environ.get("SEED_EMAIL", "admin@example.com")
        password = os.environ.get("SEED_PASSWORD", "changeme123")
        name = os.environ.get("SEED_NAME", "Project Manager")

        user = conn.execute("SELECT * FROM users WHERE email = ? COLLATE NOCASE", (email,)).fetchone()
        if user is None:
            cursor = conn.execute(
                "INSERT INTO users (email, name, password_hash, role) VALUES (?, ?, ?, ?)",
                (email, name, hash_password(password), "admin"),
            )
            conn.commit()
            user = conn.execute("SELECT * FROM users WHERE id = ?", (cursor.lastrowid,)).fetchone()
            say(f"Created account {email} (password: {password})")
        else:
            say(f"Account {email} already exists — leaving it alone")

        data = json.loads((seed_file or SEED_FILE).read_text(encoding="utf-8"))
        project = data["project"]

        if conn.execute("SELECT 1 FROM projects WHERE code = ?", (project["code"],)).fetchone():
            say(f"Project {project['code']} already seeded — nothing to do")
            return

        with conn:
            project_id = conn.execute(
                """
                INSERT INTO projects (code, name, client, description, ntp_date, duration_months,
                                      days_per_month, hours_per_month, elapsed_day_offset, status, owner_id)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    project["code"], project["name"], project["client"], project["description"],
                    project["ntp_date"], project["duration_months"], project["days_per_month"],
                    project["hours_per_month"], project.get("elapsed_day_offset", 0),
                    project.get("status", "active"), user["id"],
                ),
            ).lastrowid

            for step in default_steps():
                conn.execute(
                    """
                    INSERT INTO workflow_steps (project_id, key, name, percent, anchor, offset_days, sort_order)
                    VALUES (?, ?, ?, ?, ?, ?, ?)
                    """,
                    (project_id, step["key"], step["name"], step["percent"],
                     step["anchor"], step["offset_days"], step["sort_order"]),
                )

            trade_ids = {}
            for trade in data["trades"]:
                trade_ids[trade["key"]] = conn.execute(
                    "INSERT INTO trades (project_id, key, name, budget_hours, color, sort_order) VALUES (?, ?, ?, ?, ?, ?)",
                    (project_id, trade["key"], trade["name"], trade["budget_hours"], trade["color"], trade["sort_order"]),
                ).lastrowid

            section_ids = {}
            for section in data["sections"]:
                section_ids[section["code"]] = conn.execute(
                    "INSERT INTO sections (project_id, code, name, sort_order) VALUES (?, ?, ?, ?)",
                    (project_id, section["code"], section["name"], section["sort_order"]),
                ).lastrowid

            # The workbook holds the schedule as elapsed months since NTP; the app
            # works in real dates, so convert on the way in.
            ntp = datetime.strptime(project["ntp_date"][:10], "%Y-%m-%d").date()
            per_month = float(project["days_per_month"])

            def as_date(months: float) -> str:
                return (ntp + timedelta(days=float(months) * per_month)).isoformat()

            for task in data["tasks"]:
                start_date = as_date(task["start_month"])
                submission_date = as_date(task["finish_month"])
                # Meetings and milestones are not design submissions, so they do
                # not follow the workflow: they are tracked pro rata by time with
                # a percentage you type. The workbook marks them itself.
                is_milestone = (
                    task["finish_month"] <= task["start_month"]
                    or "(milestone)" in task["name"].lower()
                )
                tracking = "simple" if is_milestone else "workflow"

                task_id = conn.execute(
                    """
                    INSERT INTO tasks (project_id, section_id, wbs, name, weight_points,
                                       start_date, submission_date, tracking, actual_pct, remarks, sort_order)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                    """,
                    (
                        project_id, section_ids.get(task["section_code"]), task["wbs"], task["name"],
                        task["weight_points"], start_date, submission_date, tracking,
                        task["actual_pct"], task["remarks"], task["sort_order"],
                    ),
                ).lastrowid

                for key, share in task["allocations"].items():
                    if share > 0:
                        conn.execute(
                            "INSERT INTO task_allocations (task_id, trade_id, pct) VALUES (?, ?, ?)",
                            (task_id, trade_ids[key], share),
                        )

                if task["actual_pct"] > 0:
                    conn.execute(
                        """
                        INSERT INTO progress_updates
                            (task_id, project_id, user_id, previous_pct, actual_pct, note, data_date)
                        VALUES (?, ?, ?, 0, ?, 'Imported from control workbook', ?)
                        """,
                        (task_id, project_id, user["id"], task["actual_pct"], project["ntp_date"]),
                    )

        say(
            f'Seeded "{project["name"]}" with {len(data["tasks"])} deliverables '
            f'across {len(data["sections"])} sections and {len(data["trades"])} trades.'
        )
    finally:
        conn.close()
